OrbitService/ProducerSideServiceImpl.cpp

```diff
@@ -152,103 +152,104 @@ void ProducerSideServiceImpl::SendCommandsThread(
+                                                             service_state->capture_status !=
+                                                                 CaptureStatus::kCaptureFinished;
+                                                    },
+                                                    &service_state_),
+                                                kCheckExitSendCommandsThreadInterval);
         } break;
       }
-      prev_capture_status = service_state_.capture_status;
 
       service_state_mutex_.Unlock();
+      continue;
+    }
 
-      switch (prev_capture_status) {
-        case CaptureStatus::kCaptureStarted: {
-          orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse command;
-          command.mutable_start_capture_command();
-          if (!stream->Write(command)) {
-            ERROR("Sending StartCaptureCommand to CaptureEventProducer");
-            LOG("Terminating call to ReceiveCommandsAndSendEvents as Write failed");
-            // Cause Read in ReceiveEventsThread to also fail if for some reason it hasn't already.
-            context->TryCancel();
-            return;
-          }
-          LOG("Sent StartCaptureCommand to CaptureEventProducer");
-        } break;
-
-        case CaptureStatus::kCaptureStopping: {
-          orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse command;
-          command.mutable_stop_capture_command();
-          if (!stream->Write(command)) {
-            ERROR("Sending StopCaptureCommand to CaptureEventProducer");
-            LOG("Terminating call to ReceiveCommandsAndSendEvents as Write failed");
-            // Cause Read in ReceiveEventsThread to also fail if for some reason it hasn't already.
-            context->TryCancel();
-            return;
-          }
-          LOG("Sent StopCaptureCommand to CaptureEventProducer");
-        } break;
+    // service_state_.capture_status has changed compared to prev_capture_status: handle the change.
+    switch (service_state_.capture_status) {
+      case CaptureStatus::kCaptureStarted: {
+        ++service_state_.producers_remaining;
+        *all_events_sent_received = false;
+      } break;
 
-        case CaptureStatus::kCaptureFinished:
-          break;
-      }
+      case CaptureStatus::kCaptureStopping:
+        break;
 
-      continue;
+      case CaptureStatus::kCaptureFinished: {
+        *all_events_sent_received = true;
+      } break;
     }
+    prev_capture_status = service_state_.capture_status;
 
-    // Wait for service_state_.capture_status to change or for service_state->exit_requested
-    // (the next iteration will handle the change).
-    // Use a timeout to periodically check (in the next iteration)
-    // for *terminate_send_commands_thread, set by ReceiveCommandsAndSendEvents.
-    static constexpr absl::Duration kCheckExitSendCommandsThreadInterval = absl::Seconds(1);
+    service_state_mutex_.Unlock();
```

Comment:
And at this point you can use scoped lock I think: MutexLock lock (....)

Reply:
Done.